Redesign parser

11harveyj opened the pull request:
This PR redesigns how the parser function works and moves all of the heavy lifting out of the main class into seperated functions and classes for each AVR response and function area (audio, video, DSP etc.).

This has gone through a bunch of iterations since I started this, originally using reflection to select the function based on the  alpha chars of the response but this would have had issues where the value from the AVR can also contain alpha chars (such as tuner frequency / tuner preset), so a handler had to be added between. 

I need to sort a few linq issues out too before this is ready, but would appreciate any feedback.

crowbarz commented:
Ha, I was working on the HA integration side earlier today and noted a few places in the parser where the parsing was a little loose, and also zones were flagged updated whether the value was changed from the cached value or not (this impacts whether HA is flagged to poll for updated state/attributes.)

I'll take a look later on this weekend and compare notes.

11harveyj commented:
Ha brilliant, as I was sitting here this evening going through each parser I thought to myself "well hang on, this is parsing the same response but for two different fields".

There are still a couple of things I need to sort out, one of them being the `updated_zones` if the `Response` object does not contain a value in `property_name`

The first commit shouldn't be here btw, forgot to reset the branch back to head. Its part of #15 though which is ready for merge from myside, that needs to be sorted before releasing the sound mode features in the HA integration though.

crowbarz commented:
Apologies, missed that other PR, it's merged now.

crowbarz commented:
Additionally, in updating the pioneer_async integration, I noticed that in `set_channel_levels` I believe you're trying to guard against a zone not existing, I think that needs to be changed from:
```python
if self.channel_levels.get(zone).get(channel.upper()) is not None:
```
to:
```python
if self.channel_levels.get(zone, {}).get(channel.upper()) is not None:
```
as `get` returns None if the property doesn't exist, and `get` is not available with `NoneType`.

11harveyj commented on aiopioneer/parsers/const.py:
```diff
@@ -0,0 +1,33 @@
+from enum import Enum
+
+
+class Zones(Enum):
```
Yes, I did think that a short while after I started using it haha. Will move it over :)

11harveyj commented on aiopioneer/parsers/const.py:
```diff
@@ -0,0 +1,33 @@
+    Z3 = "3"
+    HDZ = "Z"
+
+class Response():
```
Yep :)

11harveyj commented on aiopioneer/pioneer_avr.py:
```diff
@@ -942,1571 +930,62 @@ def _parse_response(self, response):
+                            )
+
+                    else:
+                        try:
```
Thanks, Python is still rusty. I think this is the most I've used of it in about 8 years.

11harveyj commented on aiopioneer/pioneer_avr.py:
```diff
@@ -942,1571 +930,62 @@ def _parse_response(self, response):
+                    current_value = getattr(self, response.base_property)
+                    if response.property_name is None:
+                        #if current_value[str(p.zone)] is not p.value:
+                        current_value[response.zone] = response.value
```
Working on the comparision side of things, although I get some errors at runtime if I use the existing comparison stuff for some reason

current_value will continue to be what is currently stored in the `PioneerAVR` class. I've used `getattr` to get the property itself from passed text in the `Response` object.

11harveyj replied:
This is resolved now in commit https://github.com/crowbarz/aiopioneer/pull/20/commits/969d969427b2fd9fa3f034a2cc39964dce5cada9

11harveyj commented on aiopioneer/parsers/__init__.py:
```diff
@@ -0,0 +1,170 @@
+        return mc(raw[2:], _param)
+    if raw.startswith("is"):
+        return is_phasecontrol(raw[2:], _param)
+    if raw.startswith("vsp"):
```
Working on this.. slowly haha

11harveyj commented on aiopioneer/parsers/system.py:
```diff
@@ -0,0 +1,262 @@
+                         base_property="amp",
+                         property_name="dimmer",
+                         zone=Zones.Z1,
+                         value=raw,
```
Added dimmer modes, I don't think there is a great need to store the raw response of this one as this hasn't seemed to have changed for a very long time and appears pretty static across AVR generations. 

I've added in `DIMMER_MODES` to the main const again and referenced in line with how other responses are parsed.

11harveyj replied:
Ahh I see now, we could add an additional field in the `Response` object for that if that suits?

11harveyj replied:
Actually just realised we already capture the raw value as it is in a `Response` object under the `raw` parameter, `value` contains the parsed value (although in some cases its not always parsed, it just depends where its used).

The `_parse_response` function would just need updating so that it stores a raw value rather than a parsed value perhaps? Although might be difficult on the HA side when reading the dicts in as attributes.

I'll update the loggers so that it displays the raw and parsed value (similar to before)

crowbarz replied:
On the HA side I already need to know the mapping between raw and translated values because in the config flow I have to show the translated value, but pass the raw value to aiopioneer.

How about defining the mapping as an enum as you've done with Zone? Then both aiopioneer and the HA integration can use either, and also map back and forth between the two for free. It might also help with sets of possible values that depend on the AVR model via subclassing and/or inheritance, in the config flow I also get the bonus of getting the set of all possible values to show to the user.

11harveyj commented:
> Additionally, in updating the pioneer_async integration, I noticed that in `set_channel_levels` I believe you're trying to guard against a zone not existing, I think that needs to be changed from:
> 
> ```python
> if self.channel_levels.get(zone).get(channel.upper()) is not None:
> ```
> 
> to:
> 
> ```python
> if self.channel_levels.get(zone, {}).get(channel.upper()) is not None:
> ```
> 
> as `get` returns None if the property doesn't exist, and `get` is not available with `NoneType`.

Zone & channel, no point trying to set the channel of a speaker that doesn't exist on the AVR (supported channels are auto discovered anyway)

11harveyj commented:
the latest commits adds some parsers for model, mac, version and source names (plus a bunch more setting related ones that we didn't have before).

I know we're already hand parsing the output for model, mac, version and source names inside of `PioneerAVR`, but hopefully we can move those out and use the new parser classes from this PR instead

Also allows fixing a bug in parsing the mac address, on my AVR the existing parser inside `PioneerAVR` was returning the following
> mac_addr : SV:B7:45:E1:C5:6F
Which is incorrect as a mac address is hex only so the maximum allowed MAC by theory is FF:FF:FF:FF:FF:FF. So I've changed how it works in the new parser functions, the old one still overrides it though when the lib is first ran.

Also added a few other features, such as parsing the network standby setting which might allow us to do something about https://github.com/crowbarz/ha-pioneer_async/issues/12 without impacting functionality for AVRs that do have a network standby mode. Perhaps even adding a checkbox on the integration setup to allow aiopioneer to automatically adjust settings like this so that its most compatible with the library maybe?

And I think we can now rename sources too, not that it happens often for people, but might be handy rather than messing around with the front panel and remote, it should update the `_source_id_to_name` and `_source_name_to_id` dicts when those events happen, which in turn would update the source lists that are returned to the HA integration I think? May even be worth getting rid of these entirely as I don't think both are actually needed, but I've not looked at that in detail.

I've not added any additional queries as of yet to the main commands. But something we could enable in the future.

11harveyj commented on aiopioneer/pioneer_avr.py:
```diff
@@ -942,1571 +930,62 @@ def _parse_response(self, response):
+        parsed_response: list = process_raw(response, self._params)
+        if parsed_response is not None:
+            for response in parsed_response:
+                if response.base_property is not None:
```
Going to mark this as resolved for now, I've used null a lot more in newer commits.

`AUA` and `AUB` still need to return something (even if its essentially blank) as some overrides for the command queue a little later on requires them (snippet of code below):

```python
# Some specific overrides for the command queue
if response.response_command in ["PWR", "FN", "AUB", "AUA"]:
	# Only request these if we're not doing a full update.
	# If we are doing a full update these will be included anyway
	if (self._full_update is False) and (
		self.tone.get("1") is not None
	) and (bool(response.value)):
		self.queue_command("query_listening_mode")
		self.queue_command("query_audio_information")
		self.queue_command("query_video_information")
	# Queue a full update
	if self.tone.get("1") is None:
		self.queue_command("FULL_UPDATE")
```

11harveyj commented on aiopioneer/parsers/__init__.py:
```diff
@@ -0,0 +1,170 @@
+
+def process_raw(raw: str, _param: dict) -> list:
+    """Handler to pass the raw response to the correct parser."""
+    if raw.startswith("PWR"):
```
I've created a list of commands in `parse.py`, I wanted to extend `PIONEER_COMMANDS` more so that it only needs to go into one place, but I didn't want to start messing around with some of the core functionality of `PioneerAVR`, also would mean applying the `Zones` enum across all of `PioneerAVR` which is out of scope for this PR.

crowbarz commented:
> And I think we can now rename sources too, not that it happens often for people, but might be handy rather than messing around with the front panel and remote, it should update the _source_id_to_name and _source_name_to_id dicts when those events happen, which in turn would update the source lists that are returned to the HA integration I think?

`_source_id_to_name` is used by `get_source_name()` and it is called via the HA integration when the zone entity is flagged as updated. `__source_name_to_id` is used by the HA integration via `get_source_dict()` during entity update as well as via `select_source` in `aiopioneer` to translate the source name specified to its ID. When this is converted to use an enum then one should be able to pass in via name or ID and eliminate all the conversions back and forth.

Last week I checked in the HA side changes to expose the services for all the `aiopioneer` side functions implemented up to then, except `set_(amp|video|dsp)_settings` and the media related functions. Are any of those `aiopioneer` side functions for these updated in this PR?

I'm still working on refactoring the HA integration config flow and adding support for the new `aiopioneer` side parameters, once that is in place I will create a new release of `aiopioneer` (to pick up #15) and also the HA integration targeting that `aiopioneer` release. Hoping to get it done this weekend, doesn't look like I have as much time as you at the moment to work on this :(

11harveyj commented:
> `_source_id_to_name` is used by `get_source_name()` and it is called via the HA integration when the zone entity is flagged as updated. `__source_name_to_id` is used by the HA integration via `get_source_dict()` during entity update as well as via `select_source` in `aiopioneer` to translate the source name specified to its ID. When this is converted to use an enum then one should be able to pass in via name or ID and eliminate all the conversions back and forth.

Indeed, for the enum works I'll do that separately as this is already quite a big PR in itself.

> Last week I checked in the HA side changes to expose the services for all the `aiopioneer` side functions implemented up to then, except `set_(amp|video|dsp)_settings` and the media related functions. Are any of those `aiopioneer` side functions for these updated in this PR?

None of the side functions have been updated , with the removal of some zones from parsers there are a couple of changes that are needed on the internal bits of the code, but nothing that will affect integrations etc.

> I'm still working on refactoring the HA integration config flow and adding support for the new `aiopioneer` side parameters, once that is in place I will create a new release of `aiopioneer` (to pick up #15) and also the HA integration targeting that `aiopioneer` release. Hoping to get it done this weekend, doesn't look like I have as much time as you at the moment to work on this :(

No problem, do you think some of the parsing can be removed from `query_device_info` now that the parsers have been defined under `parsers.SystemParsers` ? I've ran it and commented out the lines of code relating to parsing with no impact so far, but was quite a limited test.

crowbarz commented:
> [Do] you think some of the parsing can be removed from query_device_info now that the parsers have been defined under `parsers.SystemParsers` ?

I'm only going to be able to test the updated parser code on my own VSX-930, it will need to be released to test this out on a wider range of AVRs.

On return from `query_device_info`, it is expected that `self.model` is set and `self._default_params` is set via `_set_default_params_model`. Which lines have you commented out to remove the parsing, and when they are commented out what is generating the `system_query_(mac_address|software_version|model)` commands to be sent for the parsers to receive responses to parse?

Note that `send_command` waits for a response with a given prefix to be returned, but it's not deterministic whether the response will already be parsed on its return - the only guarantee on non-failure is that a response with the expected prefix has been received. Both `_connection_listener` and `send_command` (via `send_raw_request`) await `_read_response`, but either could execute first once a response is actually received. It may be sufficient to `await asyncio.sleep(0)` to yield to the `_connection_listener` task which processes the response, so that the respective `mac_address`, `software_version` and/or `avr_model` have run and set the appropriate properties.

crowbarz commented:
I started up my dev HA instance with the aiopioneer version from this PR @ c6aa6c7f875ddc6b48d6f218fe5d986625ea8f95 , and on startup it continuously cycled through full updates. It shouldn't be possible to trigger a full update from within an already running full update, so not quite sure what is happening here. Logs from the middle of the cycle below.

```
2023-03-26 20:24:06.749 INFO (MainThread) [aiopioneer.pioneer_avr] updating AVR status (full=True, last updated 0.001 s ago)
2023-03-26 20:24:06.749 DEBUG (MainThread) [aiopioneer.pioneer_avr] delaying command for 0.090 s
2023-03-26 20:24:06.840 DEBUG (MainThread) [aiopioneer.pioneer_avr] sending AVR command: ?P
2023-03-26 20:24:06.850 INFO (MainThread) [aiopioneer.pioneer_avr] Zone 1: power: True (0)
2023-03-26 20:24:06.850 DEBUG (MainThread) [aiopioneer.pioneer_avr] >> PioneerAVR.queue_command(FULL_UPDATE)
2023-03-26 20:24:06.850 INFO (MainThread) [aiopioneer.pioneer_avr] Scheduling command queue. (['FULL_UPDATE'])
2023-03-26 20:24:06.850 DEBUG (MainThread) [aiopioneer.pioneer_avr] >> PioneerAVR._command_queue_schedule()
2023-03-26 20:24:06.850 DEBUG (MainThread) [aiopioneer.util] >> cancel_task(command_queue): already completed
2023-03-26 20:24:06.851 DEBUG (MainThread) [aiopioneer.pioneer_avr] >> PioneerAVR._command_queue
2023-03-26 20:24:06.851 INFO (MainThread) [aiopioneer.pioneer_avr] Command Queue Executing: FULL_UPDATE
2023-03-26 20:24:06.851 DEBUG (MainThread) [aiopioneer.pioneer_avr] AVR command ?P returned response: PWR0
2023-03-26 20:24:06.851 DEBUG (MainThread) [aiopioneer.pioneer_avr] delaying command for 0.090 s
2023-03-26 20:24:06.851 DEBUG (MainThread) [aiopioneer.pioneer_avr] Command Queue Finished.
2023-03-26 20:24:06.941 DEBUG (MainThread) [aiopioneer.pioneer_avr] sending AVR command: ?V
2023-03-26 20:24:06.960 INFO (MainThread) [aiopioneer.pioneer_avr] Zone 1: volume: 81 (081)
2023-03-26 20:24:06.960 DEBUG (MainThread) [aiopioneer.pioneer_avr] AVR command ?V returned response: VOL081
2023-03-26 20:24:06.960 DEBUG (MainThread) [aiopioneer.pioneer_avr] delaying command for 0.082 s
2023-03-26 20:24:07.049 DEBUG (MainThread) [aiopioneer.pioneer_avr] sending AVR command: ?M
2023-03-26 20:24:07.066 INFO (MainThread) [aiopioneer.pioneer_avr] Zone 1: mute: False (1)
2023-03-26 20:24:07.067 DEBUG (MainThread) [aiopioneer.pioneer_avr] AVR command ?M returned response: MUT1
2023-03-26 20:24:07.067 DEBUG (MainThread) [aiopioneer.pioneer_avr] delaying command for 0.085 s
2023-03-26 20:24:07.153 DEBUG (MainThread) [aiopioneer.pioneer_avr] sending AVR command: ?F
2023-03-26 20:24:07.169 INFO (MainThread) [aiopioneer.pioneer_avr] Zone 1: source: 06 (06)
2023-03-26 20:24:07.169 DEBUG (MainThread) [aiopioneer.pioneer_avr] >> PioneerAVR.queue_command(FULL_UPDATE)
2023-03-26 20:24:07.169 INFO (MainThread) [aiopioneer.pioneer_avr] Zone 1: media_control_mode: None (06)
2023-03-26 20:24:07.169 DEBUG (MainThread) [aiopioneer.pioneer_avr] >> PioneerAVR.queue_command(FULL_UPDATE)
2023-03-26 20:24:07.169 DEBUG (MainThread) [aiopioneer.pioneer_avr] command FULL_UPDATE already queued, skipping
2023-03-26 20:24:07.169 INFO (MainThread) [aiopioneer.pioneer_avr] Scheduling command queue. (['FULL_UPDATE'])
2023-03-26 20:24:07.169 DEBUG (MainThread) [aiopioneer.pioneer_avr] >> PioneerAVR._command_queue_schedule()
2023-03-26 20:24:07.169 DEBUG (MainThread) [aiopioneer.util] >> cancel_task(command_queue): already completed
2023-03-26 20:24:07.170 DEBUG (MainThread) [aiopioneer.pioneer_avr] >> PioneerAVR._command_queue
2023-03-26 20:24:07.170 INFO (MainThread) [aiopioneer.pioneer_avr] Command Queue Executing: FULL_UPDATE
2023-03-26 20:24:07.176 DEBUG (MainThread) [aiopioneer.pioneer_avr] AVR command ?F returned response: FN06
2023-03-26 20:24:07.176 DEBUG (MainThread) [aiopioneer.pioneer_avr] delaying command for 0.079 s
2023-03-26 20:24:07.176 DEBUG (MainThread) [aiopioneer.pioneer_avr] Command Queue Finished.
2023-03-26 20:24:07.255 DEBUG (MainThread) [aiopioneer.pioneer_avr] sending AVR command: ?ZEP
2023-03-26 20:24:07.269 INFO (MainThread) [aiopioneer.pioneer_avr] Zone Z: power: False (1)
2023-03-26 20:24:07.270 DEBUG (MainThread) [aiopioneer.pioneer_avr] AVR command ?ZEP returned response: ZEP1
2023-03-26 20:24:07.270 DEBUG (MainThread) [aiopioneer.pioneer_avr] delaying command for 0.086 s
2023-03-26 20:24:07.359 DEBUG (MainThread) [aiopioneer.pioneer_avr] sending AVR command: ?HZV
2023-03-26 20:24:07.368 INFO (MainThread) [aiopioneer.pioneer_avr] Zone Z: volume: 58 (58)
2023-03-26 20:24:07.368 DEBUG (MainThread) [aiopioneer.pioneer_avr] AVR command ?HZV returned response: XV58
2023-03-26 20:24:07.368 DEBUG (MainThread) [aiopioneer.pioneer_avr] delaying command for 0.091 s
2023-03-26 20:24:07.460 DEBUG (MainThread) [aiopioneer.pioneer_avr] sending AVR command: ?HZM
2023-03-26 20:24:07.476 INFO (MainThread) [aiopioneer.pioneer_avr] Zone Z: mute: False (1)
2023-03-26 20:24:07.476 DEBUG (MainThread) [aiopioneer.pioneer_avr] AVR command ?HZM returned response: HZMUT1
2023-03-26 20:24:07.476 DEBUG (MainThread) [aiopioneer.pioneer_avr] delaying command for 0.084 s
2023-03-26 20:24:07.561 DEBUG (MainThread) [aiopioneer.pioneer_avr] sending AVR command: ?ZEA
2023-03-26 20:24:07.574 INFO (MainThread) [aiopioneer.pioneer_avr] Zone Z: source: 06 (06)
2023-03-26 20:24:07.574 INFO (MainThread) [aiopioneer.pioneer_avr] Zone Z: media_control_mode: None (06)
2023-03-26 20:24:07.574 DEBUG (MainThread) [aiopioneer.pioneer_avr] AVR command ?ZEA returned response: ZEA06
2023-03-26 20:24:07.574 DEBUG (MainThread) [aiopioneer.pioneer_avr] calling callback for zone 1
2023-03-26 20:24:07.574 DEBUG (MainThread) [aiopioneer.pioneer_avr] calling callback for zone Z
2023-03-26 20:24:07.575 INFO (MainThread) [aiopioneer.pioneer_avr] updating AVR status (full=True, last updated 0.001 s ago)
2023-03-26 20:24:07.575 DEBUG (MainThread) [aiopioneer.pioneer_avr] delaying command for 0.087 s
2023-03-26 20:24:07.664 DEBUG (MainThread) [aiopioneer.pioneer_avr] sending AVR command: ?P
2023-03-26 20:24:07.672 INFO (MainThread) [aiopioneer.pioneer_avr] Zone 1: power: True (0)
2023-03-26 20:24:07.672 DEBUG (MainThread) [aiopioneer.pioneer_avr] >> PioneerAVR.queue_command(FULL_UPDATE)
2023-03-26 20:24:07.672 INFO (MainThread) [aiopioneer.pioneer_avr] Scheduling command queue. (['FULL_UPDATE'])
2023-03-26 20:24:07.672 DEBUG (MainThread) [aiopioneer.pioneer_avr] >> PioneerAVR._command_queue_schedule()
2023-03-26 20:24:07.672 DEBUG (MainThread) [aiopioneer.util] >> cancel_task(command_queue): already completed
2023-03-26 20:24:07.673 DEBUG (MainThread) [aiopioneer.pioneer_avr] >> PioneerAVR._command_queue
2023-03-26 20:24:07.673 INFO (MainThread) [aiopioneer.pioneer_avr] Command Queue Executing: FULL_UPDATE
2023-03-26 20:24:07.673 DEBUG (MainThread) [aiopioneer.pioneer_avr] AVR command ?P returned response: PWR0
2023-03-26 20:24:07.673 DEBUG (MainThread) [aiopioneer.pioneer_avr] delaying command for 0.091 s
2023-03-26 20:24:07.673 DEBUG (MainThread) [aiopioneer.pioneer_avr] Command Queue Finished.
2023-03-26 20:24:07.765 DEBUG (MainThread) [aiopioneer.pioneer_avr] sending AVR command: ?V
```

11harveyj commented:
> On return from `query_device_info`, it is expected that `self.model` is set and `self._default_params` is set via `_set_default_params_model`. Which lines have you commented out to remove the parsing, and when they are commented out what is generating the `system_query_(mac_address|software_version|model)` commands to be sent for the parsers to receive responses to parse?
See parsers/system.py, line 265+ for the parsing.

My limited test involved commenting out the following lines from aiopioneer.py

856-859
863-865
871-874

> I started up my dev HA instance with the aiopioneer version from this PR @ https://github.com/crowbarz/aiopioneer/commit/c6aa6c7f875ddc6b48d6f218fe5d986625ea8f95 , and on startup it continuously cycled through full updates. It shouldn't be possible to trigger a full update from within an already running full update, so not quite sure what is happening here. Logs from the middle of the cycle below.

Any chance you could upload the full logs so I can paint a 'story' as to whats going on? I've not seen that behaviour yet. I think its because of the condition from L984 onwards, they need adjusting a bit as I've added more commands since I initially did that part.

crowbarz commented:
> Any chance you could upload the full logs so I can paint a 'story' as to whats going on? I've not seen that behaviour yet. I think its because of the condition from L984 onwards, they need adjusting a bit as I've added more commands since I initially did that part.

Yep, here is the full log: https://pastebin.com/9A7CAWFK . Fairly sure it's not anything I changed on the HA side, when I switched `aiopioneer` back to the dev branch it started up properly again.

11harveyj commented:
Interesting, its not even sending the query to get tone information, I've just updated the command override section, could you try again please?

crowbarz commented:
Same result after fast-forwarding to 56065d76567c1642edea80d1806de28c14e38dd4 , log at https://pastebin.com/KWLh7BhK . I've finally got the config flow working so will try to see what is going on tomorrow (today). It may be async reordering execution unexpectedly. There are locks elsewhere in the code to try to avoid such issues.

11harveyj commented:
Ahh, I think I've just spotted why that's happening, I haven't accounted for the disable_autoquery parameter that you have set to True.

11harveyj commented:
@crowbarz - Updated, the loop should be resolved now

crowbarz commented:
It no longer loops now on my dev setup when `disable_autoquery` is enabled, but it now loops when it is disabled ...

However there's another issue with listening_mode now being a string and that is causing an exception in the integration:
```
  File "/share/config/custom_components/pioneer_async/media_player.py", line 487, in sound_mode
    return self._pioneer.listening_mode.get(self._zone, None)
AttributeError: 'str' object has no attribute 'get'
```
Let me see what is going on there and resume testing once I've figured it out.

I've cherry-picked a few changes from `dev` as `pioneer_async==0.7.3` expects `get_source_dict()` to take a zone parameter.

crowbarz commented:
Ok, I see `listening_mode` doesn't have a zone any more, will update on the HA side.

```
   ["SR", AudioParsers.listening_mode, None],
```

Actually it might not have been looping when `disable_autoquery` is turned off, I have debug level logging for aiopioneer so there are a lot more logs. With info level logging I see it has stopped, but on startup it does seem to have parsed dsp/audio/video responses three times within 10 seconds.

11harveyj commented:
> Ok, I see `listening_mode` doesn't have a zone any more, will update on the HA side.
> 
> ```
>    ["SR", AudioParsers.listening_mode, None],
> ```

Yes, I can add it back in again, but listening_mode doesn't apply on any other zone (regardless of AVR) apart from Z1 so to tidy up whats in memory a little bit, I changed it to set it as purely a string.

> Actually it might not have been looping when `disable_autoquery` is turned off, I have debug level logging for aiopioneer so there are a lot more logs. With info level logging I see it has stopped, but on startup it does seem to have parsed dsp/audio/video responses three times within 10 seconds.

This is probably where we queue extra commands to return the current state, I've had the same thing on my AVR, but I haven't had a chance to figure out where its coming from yet.

Thanks for adding those new commits in too 👍

crowbarz commented:
I remember now why I enabled `disable_autoquery` on my test instance - I only have one AVR, and my production system is running `aiopioneer 0.3.1` in parallel which has `disable_autoquery` disabled. So the production instance will trigger the autoqueries, and if my test instance didn't have `disable_autoquery` enabled it would send out another set of queries - which is why I see the autoqueries running multiple times on the startup of my test instance.

Anyway, the updated parser has been running on my test instance for a few days and it seems to work well on my setup.

Will think about the question regarding `query_device_info`. Are there any other outstanding issues left for this PR from your perspective?

11harveyj commented:
> Will think about the question regarding `query_device_info`. Are there any other outstanding issues left for this PR from your perspective?

I don't think so, I think all concerns and issues have been addressed. I'll mark this PR as ready 👍

crowbarz commented:
I've removed the parsing from `query_device_info` and `build_source_dict` and instead waited for the responder task to populate the attributes. Also I accidentally pushed 2c2c139b350ec1473839d24dd470e16821bf511c into this PR which I needed for the config flow updates on the HA integration side.

BTW are you using a formatter in your IDE? `black` is used in my environment and run automatically by VSCode on save, it decides to format things a bit differently and its mods end up being included as changes unless I tell VSCode not to format on save.

Let me know if this still runs for you and if so I'll merge it in.

11harveyj commented:
Working well here still.

Currently using the default formatter for VSCode although I don't have any of the options enabled to make any use of them however.

No worries with the accidental push, all ends up in the same place anyway ;)
